### PR TITLE
improved performance of core_number - credit to @abhinavmehndiratta

### PR DESCRIPTION
**Pull request by sbromberger**
```
julia> @benchmark core_number(g)                                                                                                             
BenchmarkTools.Trial:                                                                                                                        
  memory estimate:  1.71 GiB                                                                                                                 
  allocs estimate:  12656029                                                                                                                 
  --------------                                                                                                                             
  minimum time:     8.276 s (9.20% GC)
  median time:      8.276 s (9.20% GC)
  mean time:        8.276 s (9.20% GC)
  maximum time:     8.276 s (9.20% GC)
  --------------
  samples:          1
  evals/sample:     1
```

```
julia> @benchmark core_number4(g)                                     
BenchmarkTools.Trial:                                                 
  memory estimate:  1.04 GiB                                          
  allocs estimate:  7963418                                                                                                                  
  --------------                                                      
  minimum time:     2.022 s (4.76% GC)
  median time:      2.030 s (4.46% GC)
  mean time:        2.029 s (4.46% GC)
  maximum time:     2.034 s (4.45% GC)
  --------------
  samples:          3
  evals/sample:     1
```

```
julia> g
{500000, 10000000} directed simple UInt32 graph
```

### CODE REVIEW

File: src/degeneracy.jl [hunk 1]
```diff
@@ -39,35 +37,50 @@ julia> core_number(g)
...
-    curr_degree = 0
-    for (i, v) in enumerate(vs)
-        if degrees[v] > curr_degree
-            append!(bin_boundaries, repeat([i], (degrees[v] - curr_degree)))
-            curr_degree = degrees[v]
-        end
+    n = nv(g)
+    deg = T.(degree(g)) # degree should really return T.
+    maxdeg = maximum(deg)
+    bin = zeros(T, maxdeg+1)
```
Comment (matbesancon) [inline src/degeneracy.jl]: This makes a lot of very short names here, should we either comment on what each is and/or use some longer names for the ones used throughout the function (not the ones local to a single for loop or so

**Comment by matbesancon** [comment]
Other than the two minor things I left, this looks good, as it's just accelerating an existing algorithm without changes in apparent behaviour, it should be fairly consensual

File: src/degeneracy.jl [hunk 2]
```diff
@@ -1,5 +1,3 @@
-# Code in this file inspired by NetworkX.
```
[the remark this answers is not in the document]
Comment (sbromberger) [reply src/degeneracy.jl]: I thought I had.

File: src/degeneracy.jl (hunk 1 above)
Comment (sbromberger) [reply src/degeneracy.jl]: I'll defer to @abhinavmehndiratta to provide better variable names.

**Comment by sbromberger** [comment]
@abhinavmehndiratta is there potential for parallelizing this?

File: src/degeneracy.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (matbesancon) [reply src/degeneracy.jl]: sorry I meant as in, should this stay?

**Comment by abhinavmehndiratta** [comment]
@sbromberger 
I've added some comments, hope you find them useful !
```julia
function core_number(g::AbstractGraph{T}) where T
    has_self_loops(g) && throw(ArgumentError("graph must not have self-loops"))
    n = nv(g)    
    deg = T.(degree(g)) # this will contain core number for each vertex of graph
    maxdeg = maximum(deg) # maximum degree of a vertex in graph
    bin = zeros(T, maxdeg+1) # used for bin-sort and storing starting positions of bins
    vert = zeros(T, n) # contains the set of vertices, sorted by their degrees
    pos = zeros(T, n) # contains positions of vertices in array vert

    # count number of vertices will be in each bin
    for v = 1:n
        bin[deg[v]+1] += one(T)
    end
    # from bin sizes determine starting positions of bins in array vert 
    start = one(T)
    for d = zero(T):maxdeg
        num = bin[d+1]
        bin[d+1] = start
        start += num
    end
    # sort the vertices in increasing order of their degrees and store in array vert
    for v in vertices(g)
        pos[v] = bin[deg[v]+1]
        vert[pos[v]] = v
        bin[deg[v]+1] += one(T)
    end

    # recover starting positions of the bins
    for d = maxdeg:-1:one(T)
       bin[d+1] = bin[d]
    end
    bin[1] = one(T)

    # cores decomposition
    for i = 1:n
        v = vert[i]
        # for each neighbor u of vertex v with higher degree we have to decrease its degree and move it for one bin to the left
        for u in all_neighbors(g, v)
            if deg[u] > deg[v]
                du = deg[u]
                pu = pos[u]
                pw = bin[du+1]
                w = vert[pw]
                if u != w
                    pos[u] = pw
                    vert[pu] = w
                    pos[w] = pu
                    vert[pw] = u
                end
                bin[du+1] += one(T)
                deg[u] -= one(T)
            end
        end
    end

    return deg
end
```

**Comment by sbromberger** [comment]
Once this passes tests, given @matbesancon 's and my review, I'll merge.

File: src/degeneracy.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (sbromberger) [reply src/degeneracy.jl]: I don't think so. The code was a 1:1 transcription of NetworkX before; now it's not. There's nothing tying this to NetworkX at this point.

**Pull request merged**